### PR TITLE
[Feature] Set up Laravel Storage and Scheduler with download endpoint

### LINKED ISSUE #9685: Setup Laravel Storage, authenticated download, and file cleanup

### PR DESCRIPTION
**Pull request by petertgiles**
🤖 Resolves #9685

## 👋 Introduction

This branch sets up Laravel Storage and Laravel Scheduler with a protected download endpoint.  This branch is just prep work so it has no way to create files other than by manual shell commands.

## 🕵️ Details

As part of the setup, this branch also installs cron to drive Laravel Scheduler.  Since there's no service infrastructure in the docker container it uses the already-existing supervisord to start and keep it alive.

## 📸 Screenshot

image

### Locally

1. (optional) Bump up the token expiry to 9999 and restart the mock-auth container. https://github.com/GCTC-NTGC/gc-digital-talent/blob/7118c58a64c0b7bb7cad0839dc7af3cf646c76c0/infrastructure/conf/mock-oauth2-server.json#L6
2. Rebuild the app
3. Log in to the app and note your userId and accessToken.
4. Manually create a file for testing
`echo "hello world" > api/storage/app/user_generated/{userId}/test.txt`

- [ ] Can download file with correct access token
```bash
 curl http://localhost:8000/api/user-generated-files/test.txt  \
     -H "authorization: Bearer {accessToken}"
```
- [ ] Cannot download the file without authorization
```bash
curl http://localhost:8000/api/user-generated-files/test.txt
```
- [ ] Get a "not found" with a valid access token but for another user
```bash
 curl http://localhost:8000/api/user-generated-files/test.txt  \
     -H "authorization: Bearer {someone else's accessToken}"
```
- [ ] Can run the prune command and nothing will happen
`app:prune-user-generated-files`
- [ ] Can wait 12 hours (or just change the timeout at /api/app/Console/Commands/PruneUserGeneratedFiles.php#L39 to `>= 0`) and run the prune cammand.  The test file is deleted.

### Azure

1. Deploy the branch to the dev app service.
2. SSH in and manually create a file for testing
` echo "hello world" > /tmp/api/storage/user-generated/{userId}/test.txt`

- [ ] Can download the file from the dev app service
`GET https://dev-talentcloud.tbs-sct.gc.ca/api/user-generated-files/test.txt`
  - Since I don't have curl on my tablet I had to use Edge's "Edit and Resend" tool to manually attach the authorization header.

## 🚚 Deployment

- Add the env variable STORAGE_USER_GENERATED_FILES_PATH=/tmp/api/storage/user-generated
  - I already added it.  Just make sure IaC is deployed at least to [Release-526](https://dev.azure.com/tbs-sct/Talent%20Cloud/_releaseProgress?_a=release-pipeline-progress&releaseId=5764)
- On deploy, make sure you get the new "Laravel Scheduler cron setup successful." post-deploy step.

### CODE REVIEW

**Comment by vd1992** [comment]
Download file authorization seems to work, quite conveniently done if you just edit default user in env file
Had to create the directory to echo the file into manually though

Attempting to test the scheduler gives me this

image

File: api/.env.example [hunk 1]
```diff
@@ -102,3 +102,6 @@ FEATURE_NOTIFICATIONS=true
 
 # Laravel Scout Driver
 SCOUT_DRIVER=pgsql
+
+# Larave Storage
```
Comment (vd1992) [inline api/.env.example:106]:
```suggestion
# Laravel Storage
```

Comment (petertgiles) [reply api/.env.example]: Good catch. Updated in [fix logging](https://github.com/GCTC-NTGC/gc-digital-talent/pull/9881/commits/9798e8538a7b90bbd101fea36bd9db9fec4d785d).

**Review by esizer — APPROVED** [review]
Works great! :clap: 

I can't test azure but local is working exactly as expected :+1:

**Comment by petertgiles** [comment]
> Attempting to test the scheduler gives me this

Thanks for taking a look!  It looks like these are problems from setting up your environment with the maintenance container (so everything is owned by root) and then trying to run the worker from a regular user account.

I tried deploying and running the tests on DEV in Azure again and it went through without any trouble.  I did notice that it's a bit weird that setup.sh creates a /tmp/bootstrap directory for caching but the setup doesn't actually create the /tmp/api directory used for framework/cache and storage.  It doesn't seem to be a problem so I'm not going to mess with it now, though.

**Comment by brindasasi** [comment]
I could be doing something wrong here.. but I can't download my file which I created under my user id. 
Will take a look at it again tmo. 
Screenshot 2024-03-26 170755

**Comment by petertgiles** [comment]
> Will take a look at it again tmo.

Thanks for trying it out!  Let me know if you want to look at it together.

**Comment by brindasasi** [comment]
> > Will take a look at it again tmo.
> 
> Thanks for trying it out! Let me know if you want to look at it together.

thanks! It works nicely on Azure ! 🙌

**Review by brindasasi — APPROVED** [review]
LGTM! 👍

**Pull request merged**